Fully implement IsFileCreatedParam abstraction

DaXcess opened the pull request:
The PR that added support for Unity v2022.3.62 (#9) was merged a little too quickly and wasn't fully implemented yet, that's my fault, I should have mentioned that.

This *draft* PR will fix whatever was broken or not yet implemented in the previous PR.

> You don't have an `ApplicableToUnityVersionsSince` so any version below 2022.3.62 will fail, you also don't have another `IsFileCreatedParam` for versions startig from 2023 so those also will fail. Only 2022.3.62 will work correctly.

Actually, quite a lot of 2023 versions seem to be fine, since most of their known offsets do not include `IsFileCreated`. I am going to look for Unity 2023 versions that are not yet listed in `FunctionOffsets` for any that actually use the `IsFileCreated`. Same goes for anything below 2022.3.62, it's only broken if it actually makes use of `IsFileCreated`. Still an oversight that needs to be fixed though, so I will be going through Unity versions to see which ones also make use of `IsFileCreated`.

DaXcess committed:
Add applicable unity version for param

DaXcess commented:
Findings so far for 2022.x and 2023.x:

- Only the very latest (CVE patched) versions use `IsFileCreated`
- For all of these versions, `IsFileCreated` takes `IsFileCreatedParam`, and not `AbsolutePathString` as parameter
- Some of these Unity versions do not have debug symbols available on Unity's symbol servers. They're still bundled with the editor though, so I'm going to be adding all the offsets for the versions I test to `FunctionOffsets.cs`

Will continue looking at 2021 and earlier versions

DaXcess commented:
Some findings for 2021:

- FPTS does not distinguish between 2021.3.45f1 and 2021.3.45f2. I have changed the version detection from using `FileVersion` to `ProductVersion`, and changed the parsing a little bit. The value behind the `f` will now be stored in `Version.Revision`.
- Unity 2021.1.x and lower do not support the .netstandard 2.1 build of FPTS (not relevant to this PR, just taking a note for testing)
- Unity 2021.1.29 (CVE patched) is the first "latest minor" version not using `IsFileCreated` that I've discovered so far

For 2020:
- No use of `IsFileCreated` found in the patched versions, likely wont see them in 2019 and older as well, but I'll still go over them just to be sure

DaXcess committed:
Add bunch of offsets, finalize param

DaXcess commented:
# Conclusion

Okay, so from what it looks like, the parameter passed `IsFileCreated` has *never* been `AbsolutePathString`. From what I can see, `IsFileCreated` is only used from 2021.2.20f1 and onwards, and seemingly only in October 2025 releases (which have patched the recent CVE).

Because of this, I have moved the `IsFileCreatedParam` unity version to 2021.2.0 and up, since it's not used at all before that, and has not changed even in the latest 2023 LTS. In fact, the abstraction is not even needed at this moment, since Unity has only ever used 1 structure that has never changed (unless they changed it in Unity 6).

I have also added a bunch of offsets for all October Unity releases for Unity 2019 up to 2023, and I have individually tested each of them with a small mod that tests the functionality of FPTS.

Due to one of the patched versions being an "f2" version, I needed to make a small change to the Unity version checker so that it also parses that number as well, otherwise FPTS would see no difference between the patched and unpatched versions (which have different offsets).

@xiaoxiao921 @KingEnderBrine Does this look okay? Anything I might have missed? If everything is OK, I'll mark this PR as ready.

KingEnderBrine commented:
`IsFileCreated` definitely was used up to 2017.3.0, so just deleting `IsFileCreatedParam` for earlier versions is wrong.

KingEnderBrine commented:
Also, your version parsing fails for older versions, since they didn't include revision in the format you would expect. Like for 2019.4.16, not that old version, the `ProductVersion` is `2019.4.16.14703470`

KingEnderBrine commented:
I have all the versions of Unity that were supported prior to the latest pr installed, and I also have an easy semi-automatic way to test all of them, so, I can run them all with your changes, to make sure nothing broke.

DaXcess commented:
> `IsFileCreated` definitely was used up to 2017.3.0, so just deleting `IsFileCreatedParam` for earlier versions is wrong.

My eyes have deceived me, for some reason when I glanced over those I thought it was the `IsFileCreated` that was 0x00. I can perform additional tests on those versions if needed. I am going to assume however that those do actually take `AbsolutePathString`s for now.

Edit: Confirmed that <2017.3.0 used AbsolutePathString's. Makes sense because why would that be implemented in this project otherwise. I've re-added IsFileCreatedParam for 3.4.0 and up (just copied the version from the other v5.v0 abstraction implementations)

DaXcess commented:
> Also, your version parsing fails for older versions, since they didn't include revision in the format you would expect. Like for 2019.4.16, not that old version, the `ProductVersion` is `2019.4.16.14703470`

Hmm that's annoying, since 2021.3.45f1 and f2 are different enough to warrant new offsets. Perhaps an idea to just parse the revision that Unity provides (like in your example: 14703470) and use that to tell the difference?

DaXcess committed:
Fix param for old Unity and fix version check

DaXcess commented:
Reintroduced `IsFileCreatedParam` for older Unity versions, and fixed the version checking to use the long revision number (also updated the numbers in `FunctionOffsets.cs`)

KingEnderBrine commented:
Looks good to me. Checked some older versions and they work correctly.

xiaoxiao921 commented:
Sorry for merging a bit too soon the PR yesterday, I'm clearly not familiar with the codebase anymore and it was a mistake on my part, I didn't wanna bother KEB so I didn't ask them for review but I probably should've done that
